src/python/zquantum/core/wip/circuit/conversions/cirq_conversions.py

```diff
@@ -0,0 +1,235 @@
+    """Convert native Orquestra get to its Cirq counterpart."""
+    try:
+        cirq_gate = ORQUESTRA_TO_CIRQ_MAPPING[type(gate)]
+        if gate.params:
+            cirq_gate = cirq_gate(*gate.params)
+        return cirq_gate(*(cirq.LineQubit(qubit) for qubit in gate.qubits))
+    except KeyError:
+        raise NotImplementedError(
+            f"Cannot convert Orquestra gate {gate} to cirq. This is probably a bug, "
+            "please reach out to the Orquestra support."
+        )
+
+
+@singledispatch
+def convert_from_cirq(obj):
+    """Convert Cirq object to its closes Orquestra counterpart.
+
+    TODO: add longer description once all conversions are done.
+    """
+    raise NotImplementedError(
+        f"Conversion from cirq to Orquestra not supported for {obj}."
+    )
+
+
+@singledispatch
+def orquestra_gate_factory_from_cirq_gate(gate: cirq.Gate) -> Callable[..., Gate]:
+    """Create a function that constructs orquestra Gate based on Cirq Gate.
+
+    Args:
+          gate: Cirq gate to base the factory on.
+    Returns:
+          function of the form f(*qubits) -> Gate. For most variants of this
+          function those will be just respective classes initializers.
+    """
+    raise NotImplementedError(f"Don't know Orquestra factory for gate: {gate}.")
+
+
+@orquestra_gate_factory_from_cirq_gate.register
+def identity_gate_factory_from_cirq_identity(_gate: cirq.IdentityGate) -> Type[I]:
+    return I
+
+
+@orquestra_gate_factory_from_cirq_gate.register
+def orquestra_gate_factory_from_eigengate(gate: cirq.EigenGate) -> Callable[..., Gate]:
+    key = (type(gate), gate.global_shift, gate.exponent)
+    if key in EIGENGATE_SPECIAL_CASES:
+        return EIGENGATE_SPECIAL_CASES[key]
+    elif key[0:2] in EIGENGATE_ROTATIONS:
+        return lambda *qubits: EIGENGATE_ROTATIONS[key[0:2]](
+            *qubits, angle=exponent_to_angle(gate.exponent)
+        )
+    else:
+        raise NotImplementedError(
+            f"Conversion of arbitrary {type(gate).__name__} gate not supported yet."
+        )
+
+
+@convert_from_cirq.register
+def convert_cirq_gate_operation_to_orquestra_gate(
+    operation: cirq.GateOperation,
+) -> Gate:
+    """Convert Cirq GateOperation to native Orquestra Gate.
+
+    Note that Cirq distinguishes concept of GateOperation and Gate. The correspondence
+    between terminology in Cirq is as follows:
+
+    Cirq Gate <-> Orquestra Gate subclass
+    Cirq GateOperation <-> A concrete instance of some Orquestra Gate,
+
+    which is why this function accepts GateOperation instead of Gate.
+
+    Args:
+        operation: operation to be converted.
+    Returns:
+        Native Orquestra Gate.
+    """
+    if not all(isinstance(qubit, cirq.LineQubit) for qubit in operation.qubits):
+        raise NotImplementedError(
+            "Currently conversions from cirq to Orquestra is supported only for "
+            "gate operations with LineQubits."
+        )
```

Comment:
How come we are only supporting `LineQubit`s?

Reply:
I only implemented LineQubit because when I joined the rework efforts, we supported only `int` qubits in our gates. If we want to support `GridQubit`s too, we would need to either:
- Support other types as qubits (e.g. tuples of ints)
- Somehow linearizing `GridQubit` (e.g. by mapping qubits from a grid to their linear indices)
Both of those approaches has their challenges, and I suggest postponing it to another PR.

Reply:
Leaving this unresolved for now, just to leave the discussion visible.